source/opt/types.h

```diff
@@ -171,32 +247,45 @@ class Matrix : public Type {
   Matrix(Type* element_type, uint32_t count);
   Matrix(const Matrix&) = default;
 
-  bool IsSame(Type* that) const override;
+  bool IsSame(const Type* that) const override;
   std::string str() const override;
   const Type* element_type() const { return element_type_; }
   uint32_t element_count() const { return count_; }
 
   Matrix* AsMatrix() override { return this; }
   const Matrix* AsMatrix() const override { return this; }
 
+  void GetExtraHashWords(std::vector<uint32_t>* words) const override;
+
  private:
   Type* element_type_;
   uint32_t count_;
 };
 
 class Image : public Type {
  public:
-  Image(Type* sampled_type, SpvDim dim, uint32_t depth, uint32_t arrayed,
-        uint32_t ms, uint32_t sampled, SpvImageFormat format,
-        SpvAccessQualifier access_qualifier = SpvAccessQualifierReadOnly);
+  Image(Type* type, SpvDim dimen, uint32_t d, uint32_t array,
+        uint32_t multisample, uint32_t sampling, SpvImageFormat f,
+        SpvAccessQualifier qualifier = SpvAccessQualifierReadOnly);
   Image(const Image&) = default;
 
-  bool IsSame(Type* that) const override;
+  bool IsSame(const Type* that) const override;
   std::string str() const override;
 
   Image* AsImage() override { return this; }
   const Image* AsImage() const override { return this; }
 
+  const Type* sampled_type() const { return sampled_type_; }
+  SpvDim dim() const { return dim_; }
+  uint32_t depth() const { return depth_; }
+  uint32_t arrayed() const { return arrayed_; }
+  uint32_t ms() const { return ms_; }
+  uint32_t sampled() const { return sampled_; }
```

Comment:
I don't know what is best here, but this can have values 0, 1, or 2.  Do we want this to have an enum?  Code that uses it will be much more readable.

Reply:
I'm going to leave this as is for now at least.